File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/utils/JsonTokenizer.java
package fr.hardel.asset_editor.client.javafx.lib.utils;

import java.util.ArrayList;
import java.util.List;

public final class JsonTokenizer {

    public enum TokenType {
        STRING("json-string"),
        NUMBER("json-number"),
        BOOLEAN("json-boolean"),
        NULL("json-null"),
        PROPERTY("json-property"),
        PUNCTUATION("json-punctuation"),
        WHITESPACE(null);

        private final String highlightName;

        TokenType(String highlightName) {
            this.highlightName = highlightName;
        }

        public String highlightName() {
            return highlightName;
        }
    }

    public record Token(TokenType type, String value, int start, int end) {
    }

    public static List<Token> tokenize(String json) {
        String source = json == null ? "" : json;
        List<Token> tokens = new ArrayList<>();
        int i = 0;

        while (i < source.length()) {
            char ch = source.charAt(i);

            if (Character.isWhitespace(ch)) {
                int start = i;
                while (i < source.length() && Character.isWhitespace(source.charAt(i)))
                    i++;
                tokens.add(new Token(TokenType.WHITESPACE, source.substring(start, i), start, i));
                continue;
            }

            if (ch == '"') {
                int start = i;
                i++;
                boolean isProperty = false;

                while (i < source.length() && source.charAt(i) != '"') {
                    if (source.charAt(i) == '\\') {
                        i++;
                        if (i < source.length())
                            i++;
                    } else {
                        i++;
                    }
                }

                if (i < source.length())
                    i++;

                int j = i;
                while (j < source.length() && Character.isWhitespace(source.charAt(j)))
                    j++;
                if (j < source.length() && source.charAt(j) == ':')
                    isProperty = true;

                tokens.add(new Token(
                    isProperty ? TokenType.PROPERTY : TokenType.STRING,
                    source.substring(start, i),
                    start,
                    i
                ));
                continue;
            }

            if ("{}[]:,".indexOf(ch) >= 0) {
                tokens.add(new Token(TokenType.PUNCTUATION, String.valueOf(ch), i, i + 1));
                i++;
                continue;
            }

            if (Character.isDigit(ch) || ch == '-') {
                int start = i;
                while (i < source.length() && isNumberChar(source.charAt(i)))
                    i++;
                tokens.add(new Token(TokenType.NUMBER, source.substring(start, i), start, i));
                continue;
            }

            if (source.startsWith("true", i)) {
                tokens.add(new Token(TokenType.BOOLEAN, "true", i, i + 4));
                i += 4;
                continue;
            }

            if (source.startsWith("false", i)) {
                tokens.add(new Token(TokenType.BOOLEAN, "false", i, i + 5));
                i += 5;
                continue;
            }

            if (source.startsWith("null", i)) {
                tokens.add(new Token(TokenType.NULL, "null", i, i + 4));
                i += 4;
                continue;
            }

            tokens.add(new Token(TokenType.PUNCTUATION, String.valueOf(ch), i, i + 1));
            i++;
        }

        return List.copyOf(tokens);
    }

    private static boolean isNumberChar(char ch) {
        return Character.isDigit(ch) || ch == '.' || ch == 'e' || ch == 'E' || ch == '+' || ch == '-';
    }

    private JsonTokenizer() {
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/utils/JsonTokenizer.kt
package fr.hardel.asset_editor.client.compose.lib.utils

object JsonTokenizer {

    enum class TokenType(val highlightName: String?) {
        STRING("json-string"),
        NUMBER("json-number"),
        BOOLEAN("json-boolean"),
        NULL("json-null"),
        PROPERTY("json-property"),
        PUNCTUATION("json-punctuation"),
        WHITESPACE(null)
    }

    data class Token(
        val type: TokenType,
        val value: String,
        val start: Int,
        val end: Int
    )

    fun tokenize(json: String?): List<Token> {
        val source = json ?: ""
        val tokens = ArrayList<Token>()
        var index = 0

        while (index < source.length) {
            val ch = source[index]

            if (ch.isWhitespace()) {
                val start = index
                while (index < source.length && source[index].isWhitespace()) {
                    index++
                }
                tokens += Token(TokenType.WHITESPACE, source.substring(start, index), start, index)
                continue
            }

            if (ch == '"') {
                val start = index
                index++
                var isProperty = false

                while (index < source.length && source[index] != '"') {
                    if (source[index] == '\\') {
                        index++
                        if (index < source.length) {
                            index++
                        }
                    } else {
                        index++
                    }
                }

                if (index < source.length) {
                    index++
                }

                var lookahead = index
                while (lookahead < source.length && source[lookahead].isWhitespace()) {
                    lookahead++
                }
                if (lookahead < source.length && source[lookahead] == ':') {
                    isProperty = true
                }

                tokens += Token(
                    if (isProperty) TokenType.PROPERTY else TokenType.STRING,
                    source.substring(start, index),
                    start,
                    index
                )
                continue
            }

            if ("{}[]:,".indexOf(ch) >= 0) {
                tokens += Token(TokenType.PUNCTUATION, ch.toString(), index, index + 1)
                index++
                continue
            }

            if (ch.isDigit() || ch == '-') {
                val start = index
                while (index < source.length && isNumberChar(source[index])) {
                    index++
                }
                tokens += Token(TokenType.NUMBER, source.substring(start, index), start, index)
                continue
            }

            if (source.startsWith("true", index)) {
                tokens += Token(TokenType.BOOLEAN, "true", index, index + 4)
                index += 4
                continue
            }

            if (source.startsWith("false", index)) {
                tokens += Token(TokenType.BOOLEAN, "false", index, index + 5)
                index += 5
                continue
            }

            if (source.startsWith("null", index)) {
                tokens += Token(TokenType.NULL, "null", index, index + 4)
                index += 4
                continue
            }

            tokens += Token(TokenType.PUNCTUATION, ch.toString(), index, index + 1)
            index++
        }

        return tokens.toList()
    }

    private fun isNumberChar(ch: Char): Boolean =
        ch.isDigit() || ch == '.' || ch == 'e' || ch == 'E' || ch == '+' || ch == '-'
}
